### ISSUE TITLE
tf.argmin across longer axis significantly slower than tf.reduce_min

### ISSUE THREAD

**Issue by ctargon** [tensorflow/tensorflow#42521]
<em>Please make sure that this is an issue related to performance of TensorFlow.
As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:performance_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Amazon Linux AMI 2018.03
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: n/a
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.0,2.1,2.2
- Python version: 3.6
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 10.0, 10.1
- GPU model and memory: V100 16GB

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
When finding the argmin across axis=1 on a large, two dimensional tensor, the performance significantly degrades on GPU. We are seeing almost 10x slower on GPU rather than on CPU. Argmin is okay when taking the global argmin, however when specifying axis=1, is very slow. You can even get around it by using tf.map_fn then taking the global argmin, but that is obviously not optimal.

**Describe the expected behavior**

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.
```
import time
begin = time.time()
for _ in range(1000):
    a = tf.random.uniform(shape=(20, 150000), 
                                 minval=0, 
                                 maxval=1, 
                                 dtype=tf.float64)
    c = tf.argmin(a, axis=1)
end = time.time()
end - begin
# around 25 seconds
```

Now do not use axis=1

```
import time
begin = time.time()
for _ in range(1000):
    a = tf.random.uniform(shape=(20, 150000), 
                                 minval=0, 
                                 maxval=1, 
                                 dtype=tf.float64)
    c = tf.argmin(a)
end = time.time()
end - begin
# around 0.5 seconds
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by Saduf2019** [comment]
@ctargon 
I ran the code shared on tf-nightly please find the [gist here](https://colab.research.google.com/gist/Saduf2019/f26fa4c336701d9e60c91e5bd0cd8b1e/untitled379.ipynb) and confirm if it replicates the issue reported.

**Comment by Saduf2019** [comment]
@ctargon 
I ran the shared code on tf 2.2 gpu and do not see much time been consumed, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/def6ebade137fc9e0950dbc3894dbbd1/untitled389.ipynb)

**Comment by ctargon** [comment]
I still think there is a performance issue with the argmin function.  When I run the code, you are right there does not seem to be as much a difference between axis=1 and the global argmin, both time in around 9 seconds. 

However, if you change the function call to reduce_min, then the operation runs in ~0.15 seconds. That is over 50x faster than the argmin function, which does not seem correct. How could argmin be that much slower, when all it needs is to track the indices for the min function?

**Comment by ctargon** [comment]
Also note that i have been mistaken in saying "global argmin", rather it is the argmin for default axis=0. When the shape is (20,150000), the axis=0 means we are parallelizing over the longer axis. When axis=1, we are parallelizing over the shorter axis and thus each thread has to solve the argmin of a larger vector. This can explain why there is a difference between tf.argmin(u, axis=0) and tf.argmin(u, axis=1), but it does not explain why the reduce_min computation is over 50x faster, especially considering these are very similar ops.

**Comment by zhezherun** [comment]
It looks like in order to get precise timings, we need to make sure that all the operations are actually fully evaluated. I suspect that TF2 is doing some kind of lazy/delayed evaluation, and the operations like tf.random.uniform or tf.argmin might not be actually performed immediately.

I found that one way to force the evaluation is to print something that depends on tensor values. This is confirmed by tf.profiler: without that print statement (see below), there are fewer than 1000 GPU kernel calls to generate randoms, for example. I tried searching TF docs but couldn't find any mention of this behaviour there. Could someone confirm if my understanding about lazy evaluation is correct?

Here is a new test case that makes sure that all operations have been fully executed before we stop the timer:
```import time
begin = time.time()
c = 1
for _ in range(1000):
    a = tf.random.uniform(shape=(20, 150000), 
                                 minval=0, 
                                 maxval=1, 
                                 dtype=tf.float64)
    # Uncomment one of the 4 lines below
    #c = c + tf.math.reduce_min(a, axis=0)  # 0.26 seconds total time if this line is used
    #c = c + tf.math.reduce_min(a, axis=1)  # 0.31 seconds
    #c = c + tf.argmin(a, axis=0)  # 0.26 seconds
    c = c + tf.argmin(a, axis=1)  # 17.6 seconds!!!
print(tf.math.reduce_sum(c))  # Force evaluation
end = time.time()
print(end - begin)
```

I can see that tf.math.reduce_min is fast with either axis=0 or axis=1, but tf.argmin is very slow with axis=1.

**Comment by nikitamaia** [comment]
Hi @ctargon, can you clarify the performance issue here? As you point out the time difference in time between axis=0 and axis=1 is probably due to the difference in size of the two dimensions. And I'm not seeing much of a difference. Is the issue here about a difference in performance between argmin on CPU and GPU? Or argmin and reduce_min on GPU? Or something else?

**Comment by ctargon** [comment]
@nikitamaia If you run the example @zhezherun provides in Colab, you see that reduce_min, regardless of axis, runs in around 0.2 seconds. Similarly, you see argmin with axis=0 have the same performance. However, when argmin has axis = 1, the run time increases by almost 50x. I am not sure how you are not seeing much of a difference, as the comments in @zhezherun post regarding time appear to hold true on Colab for me as well.

How is it that two (seemingly) reduction algorithms, that should in theory have very similar performance, differ so widely? I would expect argmin across axis = 1 to have a very similar runtime compared to reduce_min across axis = 1. If argmin is implemented using reduction, then I would expect its performance to be much better.

**Comment by nikitamaia** [comment]
Thanks for clarifying. I've changed the title of the issue to reflect the comparison between argmin and reduce_min. I do see the performance difference between reduce_min and argmin. 

I don't know that argmin is necessarily implemented with reduction in the same way that reduce_min, reduce_all, reduce_any, etc are. I found a similar post on stack overflow for the numpy equivalents that might be of interest.
Is this difference a performance blocker for you? Have you tried utilizing @tf.function?

**Comment by ctargon** [comment]
It actually is a bit of a performance blocker right now, which is how we found it. We did try @tf.function, but the results are the same. We also tried different functional calls of argmin, such as tf.keras.backend.argmin, etc. but it appears they all point to the same backend function.

**Comment by ctargon** [comment]
any update?

**Comment by mohantym** [comment]
Hi @ctargon !
The drastic difference has disappeared now . Attached gist in [2.8](https://colab.sandbox.google.com/gist/mohantym/cd952cb56e95956f291c17fc39dfd1e4/git_42521.ipynb#scrollTo=WhSQ8ufOY8Hz) , [2.9](https://colab.sandbox.google.com/gist/mohantym/2d67072b6b0da5cef7bd76379520aa52/git_42521.ipynb#scrollTo=XLX3WUk4ZlD4) and [nightly ](https://colab.sandbox.google.com/gist/mohantym/311394c940ea9cb45e83cb29e87a331d/git_42521.ipynb#scrollTo=4lVKJ5TUZsJA) . Can we consider this issue as resolved now.

Thank you!
